src/accelerate/accelerator.py

```diff
@@ -1324,15 +1324,19 @@ def prepare_model(self, model: torch.nn.Module, device_placement: bool = None, e
                 model = torch.nn.parallel.DistributedDataParallel(model, **kwargs)
         if self.native_amp:
             model._original_forward = model.forward
+            model_forward_func = model.forward.__func__ if hasattr(model.forward, "__func__") else model.forward
             if self.mixed_precision == "fp16" and is_torch_version(">=", "1.10"):
-                model.forward = MethodType(torch.cuda.amp.autocast(dtype=torch.float16)(model.forward.__func__), model)
+                new_forward = torch.cuda.amp.autocast(dtype=torch.float16)(model_forward_func)
             elif self.mixed_precision == "bf16" and self.distributed_type != DistributedType.TPU:
-                model.forward = MethodType(
-                    torch.autocast(device_type=self.device.type, dtype=torch.bfloat16)(model.forward.__func__), model
-                )
+                new_forward = torch.autocast(device_type=self.device.type, dtype=torch.bfloat16)(model_forward_func)
             else:
-                model.forward = MethodType(torch.cuda.amp.autocast()(model.forward.__func__), model)
-            model.forward = MethodType(convert_outputs_to_fp32(model.forward.__func__), model)
+                new_forward = torch.cuda.amp.autocast()(model_forward_func)
+            model.forward = MethodType(new_forward, model) if hasattr(model.forward, "__func__") else new_forward
+            model.forward = (
+                MethodType(convert_outputs_to_fp32(model.forward.__func__), model)
+                if hasattr(model.forward, "__func__")
+                else convert_outputs_to_fp32(model.forward)
+            )
```

Comment:
Can we do this on `new_forward` before wrapping? Would avoid doing the test twice.

Reply:
Hello, done.